<a href="https://colab.research.google.com/github/Wilson1994/DTA-2026/blob/main/project_store_sales_analys/sales_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Анализ продаж магазина

### Входящий Dataset

In [1]:
# Запусти цей блок, щоб отримати датасет
import pandas as pd
import numpy as np

np.random.seed(42)
n = 300

categories = ['Молочне', 'Хліб та випічка', 'М\'ясо та риба', 'Овочі та фрукти', 'Напої']
products = {
    'Молочне':          ['Молоко 1л', 'Кефір 0.5л', 'Сир кисломолочний', 'Масло вершкове', 'Йогурт'],
    'Хліб та випічка':  ['Хліб білий', 'Батон', 'Булочка', 'Хліб житній', 'Круасан'],
    'М\'ясо та риба':   ['Куряче філе', 'Свинина', 'Ковбаса', 'Риба хек', 'Сосиски'],
    'Овочі та фрукти':  ['Картопля', 'Морква', 'Яблука', 'Банани', 'Помідори'],
    'Напої':            ['Вода 1.5л', 'Сік апельсиновий', 'Чай', 'Кава мелена', 'Лимонад'],
}
prices = {
    'Молоко 1л': 38, 'Кефір 0.5л': 22, 'Сир кисломолочний': 55, 'Масло вершкове': 95, 'Йогурт': 30,
    'Хліб білий': 28, 'Батон': 25, 'Булочка': 12, 'Хліб житній': 32, 'Круасан': 18,
    'Куряче філе': 165, 'Свинина': 210, 'Ковбаса': 145, 'Риба хек': 120, 'Сосиски': 98,
    'Картопля': 20, 'Морква': 15, 'Яблука': 45, 'Банани': 52, 'Помідори': 60,
    'Вода 1.5л': 18, 'Сік апельсиновий': 68, 'Чай': 85, 'Кава мелена': 155, 'Лимонад': 35,
}

chosen_cats = np.random.choice(categories, n, p=[0.25, 0.20, 0.20, 0.20, 0.15])
chosen_products = [np.random.choice(products[c]) for c in chosen_cats]

df_origin = pd.DataFrame({
    'transaction_id': range(1001, 1001 + n),
    'date':           pd.to_datetime(
                          np.random.choice(pd.date_range('2024-03-01', '2024-03-31'), n)
                      ),
    'hour':           np.random.choice(range(8, 21), n, p=[0.04,0.06,0.08,0.10,0.12,0.12,0.11,0.10,0.09,0.08,0.06,0.04,0.00]),
    'category':       chosen_cats,
    'product':        chosen_products,
    'quantity':       np.random.randint(1, 6, n),
    'price_uah':      [prices[p] for p in chosen_products],
    'customer_age':   np.random.randint(18, 70, n),
    'payment':        np.random.choice(['Готівка', 'Картка', 'Телефон'], n, p=[0.35, 0.50, 0.15]),
})


In [2]:
df = df_origin.copy()

In [3]:
df

,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment
0,1001,2024-03-17,17,Хліб та випічка,Круасан,4,18,28,Готівка
1,1002,2024-03-13,14,Напої,Кава мелена,3,155,40,Картка
2,1003,2024-03-01,18,Овочі та фрукти,Яблука,1,45,43,Картка
3,1004,2024-03-02,8,М'ясо та риба,Риба хек,3,120,51,Готівка
4,1005,2024-03-09,19,Молочне,Йогурт,5,30,63,Картка
...,...,...,...,...,...,...,...,...,...
295,1296,2024-03-08,10,М'ясо та риба,Сосиски,4,98,48,Готівка
296,1297,2024-03-05,13,Овочі та фрукти,Морква,5,15,28,Телефон
297,1298,2024-03-04,11,Молочне,Молоко 1л,3,38,27,Картка
298,1299,2024-03-24,14,М'ясо та риба,Ковбаса,1,145,20,Готівка


#### Блок 1 - Перше знайомство з даними

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   transaction_id  300 non-null    int64         
 1   date            300 non-null    datetime64[ns]
 2   hour            300 non-null    int64         
 3   category        300 non-null    object        
 4   product         300 non-null    object        
 5   quantity        300 non-null    int64         
 6   price_uah       300 non-null    int64         
 7   customer_age    300 non-null    int64         
 8   payment         300 non-null    object        
dtypes: datetime64[ns](1), int64(5), object(3)
memory usage: 21.2+ KB


In [5]:
df['total_uah'] = df.quantity * df.price_uah

In [6]:
print(df.head(7))

   transaction_id       date  hour         category         product  quantity  \
0            1001 2024-03-17    17  Хліб та випічка         Круасан         4   
1            1002 2024-03-13    14            Напої     Кава мелена         3   
2            1003 2024-03-01    18  Овочі та фрукти          Яблука         1   
3            1004 2024-03-02     8    М'ясо та риба        Риба хек         3   
4            1005 2024-03-09    19          Молочне          Йогурт         5   
5            1006 2024-03-03    13          Молочне      Кефір 0.5л         3   
6            1007 2024-03-31    15          Молочне  Масло вершкове         2   

   price_uah  customer_age  payment  total_uah  
0         18            28  Готівка         72  
1        155            40   Картка        465  
2         45            43   Картка         45  
3        120            51  Готівка        360  
4         30            63   Картка        150  
5         22            51   Картка         66  
6       

In [7]:
df.head(7) # display

,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
0,1001,2024-03-17,17,Хліб та випічка,Круасан,4,18,28,Готівка,72
1,1002,2024-03-13,14,Напої,Кава мелена,3,155,40,Картка,465
2,1003,2024-03-01,18,Овочі та фрукти,Яблука,1,45,43,Картка,45
3,1004,2024-03-02,8,М'ясо та риба,Риба хек,3,120,51,Готівка,360
4,1005,2024-03-09,19,Молочне,Йогурт,5,30,63,Картка,150
5,1006,2024-03-03,13,Молочне,Кефір 0.5л,3,22,51,Картка,66
6,1007,2024-03-31,15,Молочне,Масло вершкове,2,95,46,Телефон,190


In [8]:
df.dtypes

,0
transaction_id,int64
date,datetime64[ns]
hour,int64
category,object
product,object
quantity,int64
price_uah,int64
customer_age,int64
payment,object
total_uah,int64


2. Дізнайся розмір таблиці (рядки × стовпці).


In [9]:
print(f"Размер: {df.shape[0]} рядков х {df.shape[1]} столбцов")

Размер: 300 рядков х 10 столбцов


In [10]:
print("Тип данных:")
display(df.dtypes)

Тип данных:


,0
transaction_id,int64
date,datetime64[ns]
hour,int64
category,object
product,object
quantity,int64
price_uah,int64
customer_age,int64
payment,object
total_uah,int64


4. Перевір, чи є пропущені значення.

In [11]:
print(f"Пропущенные занчения по столлбцам:\n{df.isnull().sum()}")

Пропущенные занчения по столлбцам:
transaction_id    0
date              0
hour              0
category          0
product           0
quantity          0
price_uah         0
customer_age      0
payment           0
total_uah         0
dtype: int64


5. Виведи основну статистику (describe) для числових стовпців.

In [12]:
print("Статистика:\n")
df.describe().round(2)

Статистика:



,transaction_id,date,hour,quantity,price_uah,customer_age,total_uah
count,300.00,300,300.00,300.00,300.00,300.0,300.00
mean,1150.50,2024-03-16 02:14:24,13.72,2.97,63.45,43.0,186.97
min,1001.00,2024-03-01 00:00:00,8.00,1.00,12.00,18.0,12.00
25%,1075.75,2024-03-08 00:00:00,12.00,2.00,22.00,31.0,60.00
50%,1150.50,2024-03-16 12:00:00,14.00,3.00,38.00,42.5,120.00
75%,1225.25,2024-03-24 00:00:00,16.00,4.00,95.75,55.0,256.25
max,1300.00,2024-03-31 00:00:00,19.00,5.00,210.00,69.0,840.00
std,86.75,NaN,2.74,1.44,51.68,14.6,180.27


ВЫВОД: датасет имеет информацию про 300 транзакций, 9 столбцов, пропущенных значений нет.
Средння сумма чека 186.97 грн.
Минимальный чек - 12.00 грн. ()
Максимальный чек - 840.00 грн. ()
Подозрительных данные не найдено.


```
[1, 7, 9, 10, 12, 15]
```
Медіана: (9 + 10) / 2 = 9,5
```
[1, 7, 9, 10, 12, 15, 20]
```
Медіана: 10
---
```
[1, 7, 9, 10, 12, 15, 20]
```
Медіана: 10 - `50%`  
`min` = 1  
`max` = 20  
`count` = 7  
`mean` = 10.57  
`25%` ->  
0% - min -> 25% -> 50% - median -> 75% -> 100% - max
std ->  
10.57 - 1 = 9.57
10.57 - 7 = 3.57
..
20 - 10.57 = 9.43

Блок 2 - Фільтрація та відбір

1. Відбери всі покупки категорії «М'ясо та риба».

In [13]:
df[df['category'] == "М'ясо та риба"].head()

,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
3,1004,2024-03-02,8,М'ясо та риба,Риба хек,3,120,51,Готівка,360
8,1009,2024-03-21,15,М'ясо та риба,Ковбаса,2,145,34,Картка,290
17,1018,2024-03-06,14,М'ясо та риба,Куряче філе,3,165,69,Картка,495
20,1021,2024-03-23,9,М'ясо та риба,Куряче філе,3,165,64,Телефон,495
24,1025,2024-03-10,14,М'ясо та риба,Свинина,4,210,28,Картка,840


In [14]:
meat_fish = df[df['category'] == "М'ясо та риба"]
print(f'В категории "Мясо та риба": {len(meat_fish)} транзакции')

В категории "Мясо та риба": 61 транзакции


2. Знайди транзакції, де сума (total_uah) перевищує 500 грн.

In [19]:
big = df[df['total_uah'] > 500]
print(f'Покупки > 500: {len(big)}')
print(big[['product', 'quantity', 'total_uah']].head())
display(big[['product', 'quantity', 'total_uah']].head())

Покупки > 500: 17
        product  quantity  total_uah
24      Свинина         4        840
28  Куряче філе         4        660
81      Ковбаса         4        580
89     Риба хек         5        600
93      Свинина         3        630


,product,quantity,total_uah
24,Свинина,4,840
28,Куряче філе,4,660
81,Ковбаса,4,580
89,Риба хек,5,600
93,Свинина,3,630


3. Відбери покупки, зроблені карткою (payment == 'Картка'), кількістю більше 2 одиниць.


In [24]:
cart_mult = df[(df['payment'] == 'Картка') & (df['quantity'] > 2)]

print("Карта + количество > 2", len(cart_mult), "\n")

cart_mult.head()

Карта + количество > 2 92 



,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
1,1002,2024-03-13,14,Напої,Кава мелена,3,155,40,Картка,465
4,1005,2024-03-09,19,Молочне,Йогурт,5,30,63,Картка,150
5,1006,2024-03-03,13,Молочне,Кефір 0.5л,3,22,51,Картка,66
10,1011,2024-03-23,13,Молочне,Сир кисломолочний,5,55,30,Картка,275
17,1018,2024-03-06,14,М'ясо та риба,Куряче філе,3,165,69,Картка,495


4. Знайди всі покупки покупців молодше 25 років.

In [25]:
young = df[df['customer_age'] < 25]
print("Покупатели до 25: ", len(young), '\n')
young.head()

Покупатели до 25:  36 



,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
7,1008,2024-03-01,17,Напої,Сік апельсиновий,1,68,21,Готівка,68
19,1020,2024-03-21,16,Хліб та випічка,Круасан,1,18,19,Готівка,18
26,1027,2024-03-19,13,Молочне,Йогурт,3,30,20,Готівка,90
36,1037,2024-03-04,14,Хліб та випічка,Круасан,1,18,23,Телефон,18
38,1039,2024-03-01,10,Овочі та фрукти,Помідори,1,60,22,Готівка,60


5. Скільки унікальних товарів є в датасеті?

In [27]:
n_unique = df['product'].nunique()
# n_unique

pr_unique = df['product'].unique()
# pr_unique

print("Количество уникальных товаров", n_unique, "\n")
print("Перечень уникальных товаров:\n", pr_unique)


Количество уникальных товаров 25 

Перечень уникальных товаров:
 [np.str_('Круасан') np.str_('Кава мелена') np.str_('Яблука')
 np.str_('Риба хек') np.str_('Йогурт') np.str_('Кефір 0.5л')
 np.str_('Масло вершкове') np.str_('Сік апельсиновий') np.str_('Ковбаса')
 np.str_('Картопля') np.str_('Сир кисломолочний') np.str_('Морква')
 np.str_('Куряче філе') np.str_('Хліб житній') np.str_('Батон')
 np.str_('Хліб білий') np.str_('Свинина') np.str_('Сосиски')
 np.str_('Молоко 1л') np.str_('Помідори') np.str_('Лимонад')
 np.str_('Булочка') np.str_('Банани') np.str_('Вода 1.5л') np.str_('Чай')]


Вывод: приблизительно половина покупок оплачена карткой с которых треть - больше двоих товаров в чеке.
5.7 % транзакций превышают чек в размере 500 грн., большинство сделано в категории мясо и рыба.
12% людей (покупателелй) моложе возраста (до 25 лет)

Блок 3 - Групування та агрегація

1. Підрахуй загальний виторг по кожній категорії. Відсортуй за спаданням.


In [31]:
#revenue_by_cat =
df.groupby('category')['total_uah'].sum().sort_values(ascending=False)

,total_uah
category,
М'ясо та риба,25223
Напої,10389
Молочне,10075
Овочі та фрукти,6270
Хліб та випічка,4134


In [45]:
df.groupby('category').agg({
    'total_uah': 'sum',
    'transaction_id': 'count',
    'quantity' : 'sum',
}).sort_values(by=["total_uah", "transaction_id"], ascending=False)

,total_uah,transaction_id,quantity
category,,,
М'ясо та риба,25223,61,186
Напої,10389,48,129
Молочне,10075,80,242
Овочі та фрукти,6270,55,158
Хліб та випічка,4134,56,177


2. Знайди топ-5 товарів за кількістю проданих одиниць (сума quantity).